packages/astro/src/runtime/server/transition.ts

```diff
@@ -0,0 +1,155 @@
+import type {
+	SSRResult,
+	TransitionAnimation,
+	TransitionDirectionalAnimations,
+	TransitionAnimationValue,
+} from '../../@types/astro';
+import { markHTMLString } from './escape.js';
+import { slide, fade } from '../../transitions/index.js';
+
+const transitionNameMap = new WeakMap<SSRResult, number>();
+function incrementTransitionNumber(result: SSRResult) {
+	let num = 1;
+	if(transitionNameMap.has(result)) {
+		num = transitionNameMap.get(result)! + 1;
+	}
+	transitionNameMap.set(result, num);
+	return num;
+}
+
+function createTransitionScope(result: SSRResult, hash: string) {
+	const num = incrementTransitionNumber(result);
+	return `astro-${hash}-${num}`;
+}
+export function renderTransition(result: SSRResult, hash: string, animationName: TransitionAnimationValue | undefined, transitionName: string) {
+	let animations: TransitionDirectionalAnimations | null = null;
+	switch(animationName) {
+		case 'fade': {
+			animations = fade();
+			break;
+		}
+		case 'slide': {
+			animations = slide();
+			break;
+		}
```

Comment:
Note: I was thinking if this not being tree-shakable would be an issue, but our client directives are also not treeshakable so I think this is also fine for now.